streams/src/main/java/org/apache/kafka/streams/state/internals/ListValueStore.java

```diff
@@ -0,0 +1,165 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.Serde;
+
+/**
+ * A wrapper key-value store that serializes the record values bytes as a list.
+ * As a result put calls would be interpreted as a get-append-put to the underlying RocksDB store.
+ * A put(k,null) will still delete the key, ie, the full list of all values of this key.
+ * Range iterators would also flatten the value lists and return the values one-by-one.
+ *
+ * This store is used for cases where we do not want to de-duplicate values of the same keys but want to retain all such values.
+ */
+@SuppressWarnings("unchecked")
+public class ListValueStore
+    extends WrappedStateStore<KeyValueStore<Bytes, byte[]>, Bytes, byte[]>
+    implements KeyValueStore<Bytes, byte[]> {
+
+    static private final Serde<List<byte[]>> LIST_SERDE = Serdes.ListSerde(ArrayList.class, Serdes.ByteArray());
+
+    ListValueStore(final KeyValueStore<Bytes, byte[]> bytesStore) {
+        super(bytesStore);
+    }
+
+    @Override
+    public void put(final Bytes key, final byte[] addedValue) {
+        // if the value is null we can skip the get and blind delete
+        if (addedValue == null) {
+            wrapped().put(key, null);
+        } else {
+            final byte[] oldValue = wrapped().get(key);
+
+            if (oldValue == null) {
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, Collections.singletonList(addedValue)));
+            } else {
+                final List<byte[]> list = LIST_SERDE.deserializer().deserialize(null, oldValue);
+                list.add(addedValue);
+
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, list));
+            }
+        }
+    }
+
+    @Override
+    public byte[] putIfAbsent(final Bytes key, final byte[] addedValue) {
+        final byte[] oldValue = wrapped().get(key);
+
+        if (oldValue != null) {
+            // if the value is null we can skip the get and blind delete
+            if (addedValue == null) {
+                wrapped().put(key, null);
+            } else {
+                final List<byte[]> list = LIST_SERDE.deserializer().deserialize(null, oldValue);
+                list.add(addedValue);
+
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, list));
+            }
+        }
+
+        return oldValue;
```

Comment:
Is it save to return `oldValue`? It would be a serialized `List` but the outer metered store expect a plain value? As we don't really check the result of `putIfAbsent` we could return `null`?

Reply:
In the ChangeloggingListValueStore we always return null (added a TODO as well explaining this is a hack).

Reply:
But users could disable changelogging, right?

Reply:
The only caller of the `ListValueStore.get()` is the `ChangeloggingListValueStore`. So:

* If logging is enabled, then we need to make sure that the correct value is returned from `get()` in order to log.
* If logging is disabled, then there's no caller of that `get()` anyways so there's no concern of deserialization errors.

Reply:
Yes, but this is not `get()` but `pufIfAbsent()`, right? And `putIfAbsent()` is called by the join-Processor. If we have the logging layer and the logging layer returns `null` for `putIfAbsent()` the metered store is "guarded" -- but if logging is disabled, the metered store would crash when the join-processor calls `putIfAbsent(key,null)` ?

Reply:
Ah yes, you're right. Hmm, I guess I'd have to duplicate the logic of putifabsent in both layers then.

I will file a follow-up PR.